### ISSUE TITLE
gradle export tasks misses dependencies

### ISSUE THREAD

**Issue by mnlipp** [bndtools/bnd#2392]
If a project's `launch.bndrun` has bundles in its `-runbundles` that are not also in `bnd.bnd`'s `-buildpath`, they are not (necessarily) included in the jar built by the gradle `export` task.

[Here's](https://github.com/mnlipp/test) the minimal project for reproducing the problem.

Bundle `test.bundle1` does not need `test.bundle2` when being built (imagine `test.bundle2` to be some service provider with an API provided by a third bundle). But both `test.bundle1` and `test.bundle2` are required to get a working application, as is specified by `test.bundle1/launch.bndrun` which has both bundles in its `-runbundles`. Launching this from within bndtools works, both bundles are made available.

Running `./gradew :test.bundle1:export` in a freshly checked out project (or running `./gradew clean :test.bundle1:export`) should produce a `launch.jar` (`test.bundle1/generated/distributions/executable/launch.jar`) that includes both `jar/test.bundle1.jar` *and* `jar/test.bundle2.jar`. However, it does not. The generated `launch.jar` contains only `jar/test.bundle1.jar`.

There are two workaround. Explicitly adding the `build` "task" when invoking gradle. So `./gradew clean build :test.bundle1:export` produces a `launch.jar` that contains `jar/test.bundle2.jar`. However, while this behavior can be specified on the command line, I failed to achieve the behavior with task dependencies. `export` is not my "ultimate" target, it just something that another task depends on, and you cannot depend on `build` and `:test.bundle1:export` (you can specify it in `build.gradle`, but it doesn't produce the same result).

The other work-around is to put `test.bundle2` on the `-buildpath` in `test.bundle1/bnd.bnd`. Of course, the drawback here is that now `test.bundle2`'s exported packages are available when editing the sources of `test.bundle1` and may unintentionally be included.

To summarize: it seems that the `export` task uses dependencies as can be derived from the `-buildpath` in `bnd.bnd`. But it should use the dependencies that can be derived from the `-runbundles` of the `*.bndrun` that it is processing.

**Comment by bjhargrave** [comment]
Yes, `-runbundles` is not processes for gradle dependencies since the elements of runbundles may come from many sources including unbuilt projects. This is true also for other `-run*` properties. `-runbundles` needs to be evaluated at execution time since they can otherwise create dependency cycles (this was observed during my original work on creating the gradle plugin for the Bnd workspace model). If you know your `-runbundles` refers to projects which are not already dependency via `-buildpath`, then use `-dependon` to add those other projects as dependencies to your project.

So in your example, you need `-dependon: test.bundle2`.

**Comment by mnlipp** [comment]
Thanks for the fast reply. Now that I know about it, I did find it in the bnd command reference. Though, a small hint in the description of the [gradle tasks](https://github.com/bndtools/bnd/blob/master/biz.aQute.bnd.gradle/README.md#additional-tasks) might be helpful. It *is* a bit surprising that the configuration that works fine for a launch from bndtools doesn't work for an export from gradle.
